imbi/automations/__init__.py

```diff
@@ -1,2 +1,206 @@
+        self.application = application
+        self.logger = logging.getLogger(__name__).getChild('AutomationContext')
+        self.notes: list[tuple[datetime.datetime, str]] = []
+        self.user = user
+        self.run_query = query
+
+        self._cleanups: list[CompensatingAction] = []
+        self._memory: dict[object, object] = {}
+
+    def note_progress(self, message_format: str, *args: object) -> None:
+        """Add a note to the list of notes
+
+        These could be made available in an API response so don't
+        log anything sensitive.
+        """
+        self.logger.info(message_format, *args)
+        self.notes.append((datetime.datetime.now(datetime.timezone.utc),
+                           message_format % args if args else message_format))
+
+    async def run_automation(self, automation: imbi.models.Automation,
+                             *args: object, **kwargs: object) -> None:
+        self.logger.debug('running automation %s = %r with args %r %r',
+                          automation.slug, automation.callable, args, kwargs)
+        try:
+            result = automation.callable(self, automation, *args, **kwargs)
+            if inspect.isawaitable(result):
+                await result
+        except Exception as error:
+            self.logger.exception(
+                'automation %s failed with %r, running %s cleanups',
+                automation.slug, error, len(self._cleanups))
+            await self._run_cleanups(error)
+            raise error
+
+    def add_callback(self, compensating_action: CompensatingAction) -> None:
+        """Add a cleanup action"""
+        self._cleanups.append(compensating_action)
+
+    # Implement MutableMapping interface for state storage
+    def __getitem__(self, key: object) -> object:
+        return self._memory[key]
+
+    def __delitem__(self, key: object) -> None:
+        del self._memory[key]
+
+    def __setitem__(self, key: object, value: object) -> None:
+        self._memory[key] = value
+
+    def __contains__(self, key: object) -> bool:
+        return key in self._memory
+
+    # Implement an async context manager that runs cleanups if an
+    # exception is raised
+    async def __aenter__(self) -> typing.Self:
+        return self
+
+    async def __aexit__(self, _exc_type: typing.Type[BaseException] | None,
+                        exc_val: BaseException | None,
+                        _exc_tb: types.TracebackType | None) -> None:
+        if exc_val is not None:
+            await self._run_cleanups(exc_val)
+
+    async def _run_cleanups(self, exc: BaseException) -> None:
+        cleanups, self._cleanups = self._cleanups, []
+        for cleanup in reversed(cleanups):
+            try:
+                maybe_coro = cleanup(self, exc)
+                if inspect.isawaitable(maybe_coro):
+                    await maybe_coro
+            except Exception as error:
+                self.logger.exception('cleanup %r failed with %s', cleanup,
```

Comment:
I wonder if we should let the user know that a cleanup has failed. Or maybe that's too remotely corner-case to care, and we just have them manually fix it.

Reply:
I'd lean towards the latter at this stage. If we improve the feedback (eg, SSE) then we can add it there.